### PR TITLE
driver: Port80: npcx: defer Port80 code sending to workqueue thread

### PR DESCRIPTION
**Pull request by ChiHuaL**
If the host sends Port80 postcodes frequently while EC is busy handling other tasks, the Port80 FIFO (16-byte depth) might overflow easily, especially when the host sends the postcode with the 4-byte format. This change defers the handling and sending (to the upper layer) postcodes to the system workqueue thread. It can reduce a lot of (but not all) the overflow case. Also in practice, we usually care about the latest postcodes. The older codes are not significant to the developer. This commit also lowers the printing of the overflow warning to LOG_DEBUG.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

### CODE REVIEW

File: drivers/espi/host_subs_npcx.c [hunk 1]
```diff
@@ -128,6 +128,14 @@
 #include <zephyr/irq.h>
 LOG_MODULE_REGISTER(host_sub_npcx, LOG_LEVEL_ERR);
 
+/** Minimal ring buffer implementation */
+struct ring_buf {
+	uint32_t *buf;
+	uint16_t len;
+	uint16_t head;
+	uint16_t tail;
+};
```
Comment (fabiobaltieri) [inline drivers/espi/host_subs_npcx.c:137]: Could you use the Zephyr ring buffer implementation instead? https://docs.zephyrproject.org/latest/kernel/data_structures/ring_buffers.html

Comment (keith-zephyr) [reply drivers/espi/host_subs_npcx.c:137]: +1 for using the Zephyr ring buffer.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/espi/host_subs_npcx.c (hunk 1 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: Done.

File: drivers/espi/Kconfig.npcx [hunk 2]
```diff
@@ -48,6 +48,14 @@ config ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_MULTI_BYTE
 	  EC can accept 1/2/4 bytes of Port 80 data written from the Host in an
 	  eSPI transaction.
 
+config ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_RING_BUF_SIZE
+	int "Debug Port80 ring buffer size"
+	depends on SOC_SERIES_NPCX9 && ESPI_PERIPHERAL_DEBUG_PORT_80
```
Comment (fabiobaltieri) [inline drivers/espi/Kconfig.npcx:53]: I think that this depends and the one above should match the build config that enables the actual driver to be built + the condition in the file, so that would be `depends on ESPI_NPCX && ESPI_PERIPHERAL_DEBUG_PORT_80`, which means you could move it under the "if ESPI_NPCX" below and keep this a bit simpler.

File: drivers/espi/host_subs_npcx.c [hunk 3]
```diff
@@ -148,6 +149,18 @@ struct host_sub_npcx_data {
...
+	struct ring_buf port80_ring_buf;
+	uint8_t port80_data[CONFIG_ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_RING_BUF_SIZE];
+	struct k_work work;
+#endif
+};
+
+struct npcx_dp80_buf {
+	union {
+		uint16_t u16;
+		uint8_t u8[2];
```
Comment (fabiobaltieri) [inline drivers/espi/host_subs_npcx.c:162]: Pick meaningful names for the fields. Browsing the code I see `struct can_frame` using `data` and `data_32` for a similar case, something like that would work.

File: drivers/espi/host_subs_npcx.c [hunk 4]
```diff
@@ -474,6 +487,40 @@ static void host_pmch_ibf_isr(const void *arg)
...
+	struct ring_buf *rbuf = &data->port80_ring_buf;
+	struct espi_event evt = {ESPI_BUS_PERIPHERAL_NOTIFICATION,
+				 (ESPI_PERIPHERAL_INDEX_0 << 16) | ESPI_PERIPHERAL_DEBUG_PORT80,
+				 ESPI_PERIPHERAL_NODATA};
+
+	while (!ring_buf_is_empty(rbuf)) {
+		struct npcx_dp80_buf dp80_buf;
+		uint8_t offset;
+
+		ring_buf_get(rbuf, &dp80_buf.u8[0], 2);
```
Comment (fabiobaltieri) [inline drivers/espi/host_subs_npcx.c:503]: Can you refer to the field sizeof itself instead of hardcoding the size? Maybe since you have the union you could do ` ring_buf_get(rbuf, &dp80_buf.u16, sizeof(dp80_buf.u16));`. Same in the peek below.

File: drivers/espi/host_subs_npcx.c [hunk 5]
```diff
@@ -474,6 +487,40 @@ static void host_pmch_ibf_isr(const void *arg)
...
+		offset = dp80_buf.u8[1];
+		code |= dp80_buf.u8[0] << (8 * offset);
+		if (ring_buf_is_empty(rbuf)) {
+
```
Comment (fabiobaltieri) [inline drivers/espi/host_subs_npcx.c:507]: nit: drop the blank line

File: drivers/espi/host_subs_npcx.c [hunk 6]
```diff
@@ -496,55 +543,31 @@ static void host_port80_isr(const void *arg)
...
+			ring_buf_put(rbuf, &dp80_buf.u8[0], 2);
 		}
+		k_work_submit(&host_sub_data.work);
 	}
 
 	/* If FIFO is overflow, show error msg */
 	if (IS_BIT_SET(status, NPCX_DP80STS_FOR)) {
 		inst_shm->DP80STS |= BIT(NPCX_DP80STS_FOR);
-		LOG_ERR("Port80 FIFO Overflow!");
+		LOG_DBG("Port80 FIFO Overflow!");
```
Comment (fabiobaltieri) [inline drivers/espi/host_subs_npcx.c:560]: Is this change intentional? LOG_ERR sounds reasonable here.

File: drivers/espi/Kconfig.npcx (hunk 2 above)
Comment (ChiHuaL) [reply drivers/espi/Kconfig.npcx]: `ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_MULTI_BYTE` is only supported in the NPCX9 chip series.
`ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_RING_BUF_SIZE` is used when `ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_MULTI_BYTE` is defined.
In the latest patch, I change `ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_RING_BUF_SIZE` to depend on `ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_MULTI_BYTE`.
I also change some codes related to the multiple support to be guarded by 'CONFIG_ESPI_NPCX_PERIPHERAL_DEBUG_PORT_80_MULTI_BYTE。

File: drivers/espi/host_subs_npcx.c (hunk 3 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: done.

File: drivers/espi/host_subs_npcx.c (hunk 4 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: Done.

File: drivers/espi/host_subs_npcx.c (hunk 5 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: Done

File: drivers/espi/host_subs_npcx.c (hunk 6 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c:558]: Yes, this is the intentional change. Please refer to the comment 19 and 20 in issue [b:259971621](https://partnerissuetracker.corp.google.com/issues/259971621) for the explanation,

File: drivers/espi/host_subs_npcx.c [hunk 7]
```diff
@@ -474,77 +487,85 @@ static void host_pmch_ibf_isr(const void *arg)
...
+	struct ring_buf *rbuf = &data->port80_ring_buf;
+	struct espi_event evt = {ESPI_BUS_PERIPHERAL_NOTIFICATION,
+				 (ESPI_PERIPHERAL_INDEX_0 << 16) | ESPI_PERIPHERAL_DEBUG_PORT80,
+				 ESPI_PERIPHERAL_NODATA};
+
+	while (!ring_buf_is_empty(rbuf)) {
+		struct npcx_dp80_buf dp80_buf;
+		uint8_t offset;
+
+		ring_buf_get(rbuf, &dp80_buf.offset_code[0], sizeof(dp80_buf.offset_code_16));
```
Comment (fabiobaltieri) [inline drivers/espi/host_subs_npcx.c:504]: Can you use `&dp80_buf.offset_code_16` as the *data argument? That way you have the same field as dereference and sizeof, makes it more obvious that it's the correct size you are taking.

**Comment by albertofloyd** [comment]
@rgundi

File: drivers/espi/host_subs_npcx.c (hunk 7 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: The `ring_buf_get()` function expects the 2nd argument to be 'uint8_t *. Passing `&dp80_buf.offset_code_16` to it (without casting)  causes the compiler to complains of it.
So I changed that to ` ring_buf_peek(rbuf, &dp80_buf.offset_code[0], ARRAY_SIZE(dp80_buf.offset_code));` in the updated patch.
Is that change ok with you?

Comment (fabiobaltieri) [reply drivers/espi/host_subs_npcx.c]: Got it, no `ARRAY_SIZE` is number of elements so it should definitely not be there, use `sizeof(dp80_buf.offset_code)` instead, that returns the size of the whole array.

File: drivers/espi/host_subs_npcx.c [hunk 8]
```diff
@@ -474,77 +487,85 @@ static void host_pmch_ibf_isr(const void *arg)
...
-
-	} else {
-		uint16_t port80_buf[16];
-		uint8_t count = 0;
-		uint32_t code = 0;
-
-		while (IS_BIT_SET(inst_shm->DP80STS, NPCX_DP80STS_FNE) &&
-		       count < ARRAY_SIZE(port80_buf)) {
-			port80_buf[count++] = inst_shm->DP80BUF;
+		ring_buf_peek(rbuf, &dp80_buf.offset_code[0], ARRAY_SIZE(dp80_buf.offset_code));
```
Comment (keith-zephyr) [inline drivers/espi/host_subs_npcx.c:513]: Include the comment about detecting the start of the next port 80 code.

File: drivers/espi/host_subs_npcx.c (hunk 7 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: Done.

File: drivers/espi/host_subs_npcx.c (hunk 8 above)
Comment (ChiHuaL) [reply drivers/espi/host_subs_npcx.c]: Done.

**Review by fabiobaltieri — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]

**Comment by ChiHuaL** [comment]
@albertofloyd 
Hi, 
Do you have further comment on this PR?

**Review by albertofloyd — APPROVED** [review]
